pkg/protocol/tipmanager/v1/tipmanager.go

```diff
@@ -0,0 +1,363 @@
+			rootBlocks := t.retrieveRootBlocks()
+
+			rootBlockCount := len(rootBlocks)
+			if rootBlockCount > t.optMaxStrongParents {
+				rootBlockCount = t.optMaxStrongParents
+			}
+
+			references[model.StrongParentType] = rootBlocks[:rootBlockCount]
+		}
+
+		return nil
+	})
+
+	return references
+}
+
+// StrongTipSet returns the strong tip set of the TipManager.
+func (t *TipManager) StrongTipSet() []*blocks.Block {
+	var tipSet []*blocks.Block
+	t.strongTipSet.ForEach(func(_ iotago.BlockID, tipMetadata *TipMetadata) bool {
+		tipSet = append(tipSet, tipMetadata.Block())
+
+		return true
+	})
+
+	return tipSet
+}
+
+// WeakTipSet returns the weak tip set of the TipManager.
+func (t *TipManager) WeakTipSet() (tipSet []*blocks.Block) {
+	t.weakTipSet.ForEach(func(_ iotago.BlockID, tipMetadata *TipMetadata) bool {
+		tipSet = append(tipSet, tipMetadata.Block())
+
+		return true
+	})
+
+	return tipSet
+}
+
+// Evict evicts a slot from the TipManager.
+func (t *TipManager) Evict(slotIndex iotago.SlotIndex) {
+	if t.markSlotAsEvicted(slotIndex) {
+		if evictedObjects, deleted := t.tipMetadataStorage.DeleteAndReturn(slotIndex); deleted {
+			evictedObjects.ForEach(func(_ iotago.BlockID, tipMetadata *TipMetadata) bool {
+				tipMetadata.evicted.Trigger()
+
+				return true
+			})
+		}
+	}
+}
+
+// Events returns the events of the TipManager.
+func (t *TipManager) Events() *tipmanager.Events {
+	return t.events
+}
+
+// Shutdown marks the TipManager as shutdown.
+func (t *TipManager) Shutdown() {
+	t.TriggerStopped()
+}
+
+// setupBlockMetadata sets up the behavior of the given Block.
+func (t *TipManager) setupBlockMetadata(tipMetadata *TipMetadata) {
+	tipMetadata.stronglyConnectedToTips.OnUpdate(func(_, isConnected bool) {
+		t.updateParents(tipMetadata, propagateConnectedChildren(isConnected, true))
+	})
+
+	tipMetadata.weaklyConnectedToTips.OnUpdate(func(_, isConnected bool) {
+		t.updateParents(tipMetadata, propagateConnectedChildren(isConnected, false))
+	})
+
+	tipMetadata.OnIsStrongTipUpdated(func(isStrongTip bool) {
+		if isStrongTip {
+			t.strongTipSet.Set(tipMetadata.Block().ID(), tipMetadata)
+		} else {
+			t.strongTipSet.Delete(tipMetadata.Block().ID())
+		}
+	})
+
+	tipMetadata.OnIsWeakTipUpdated(func(isWeakTip bool) {
+		if isWeakTip {
+			t.weakTipSet.Set(tipMetadata.Block().ID(), tipMetadata)
+		} else {
+			t.weakTipSet.Delete(tipMetadata.Block().ID())
+		}
+	})
+
+	tipMetadata.OnEvicted(func() {
+		t.strongTipSet.Delete(tipMetadata.Block().ID())
+		t.weakTipSet.Delete(tipMetadata.Block().ID())
+	})
+
+	tipMetadata.setTipPool(t.determineTipPool(tipMetadata))
+
+	t.events.BlockAdded.Trigger(tipMetadata)
+}
+
+// determineTipPool determines the initial TipPool of the given Block.
+func (t *TipManager) determineTipPool(tipMetadata *TipMetadata, minPool ...tipmanager.TipPool) tipmanager.TipPool {
+	if lo.First(minPool) <= tipmanager.StrongTipPool && !t.conflictDAG.AcceptanceState(tipMetadata.Block().ConflictIDs()).IsRejected() {
```

Comment:
It determines the minimum Pool you want the function to return.

If you want to downgrade a tip from the StrongTipPool then you can provide WeaKTipPool as an optional parameter and it will force the function to return a pool of type WeakTipPool or worse (possibly Dropped).

And yes it is used for downgrades when we can't pick up a tip.

Reply:
Ah, I didn't notice that StrongTipPool is smaller than WeakTipPool. I thought it's the other way.